pkg/querier/queryrange/limits.go

```diff
@@ -178,6 +186,83 @@ func (l limitsMiddleware) Do(ctx context.Context, r queryrangebase.Request) (que
 	return l.next.Do(ctx, r)
 }
 
+type querySizeLimiter struct {
+	next              queryrangebase.Handler
+	maxQueryBytesRead func(string) int
+	errorTemplate     string
+}
+
+// NewQuerySizeLimiterMiddleware creates a new Middleware that enforces query size limits.
+// The errorTemplate should format two strings: the bytes that would be read and the bytes limit.
+func NewQuerySizeLimiterMiddleware(maxQueryBytesRead func(string) int, errorTemplate string) queryrangebase.Middleware {
+	return queryrangebase.MiddlewareFunc(func(next queryrangebase.Handler) queryrangebase.Handler {
+		return &querySizeLimiter{
+			next:              next,
+			maxQueryBytesRead: maxQueryBytesRead,
+			errorTemplate:     errorTemplate,
+		}
+	})
+}
+
+// getIndexStatsForRequest return the index stats for the matchers in r's query
+func (q *querySizeLimiter) getIndexStatsForRequest(ctx context.Context, r queryrangebase.Request) (*logproto.IndexStatsResponse, error) {
+	matchers, err := syntax.ExtractMatchersFromQuery(r.GetQuery())
+	if err != nil {
+		return nil, err
+	}
+
+	// Get Stats for this query
+	var indexStatsReq definitions.Request = &logproto.IndexStatsRequest{}
```

Comment:
> We can also just target the bytes returned by the index, which seems more straightforward.

That's what the IndexStatsRequest returns, doesn't it? If so, my comment above still applies: IIRC the IndexStatsResponse can return way more bytes than what is actually in the chunks.

Reply:
The index will return the bytes in the chunks, but we may not need to query them all (in the case of reaching the `limit` first). There is some duplication due to `replication_factor` (which still needs to be deduped at query time) and some chunks being recorded in multiple indices (which are time-sharded by day). I don't think this really matters though. We'll just set the expected limit based on what the index returns rather than the amount of data that ends up being queried, which is harder to measure/not known before we run the query (in the case of `limited` queries).

For instance, if the `IndexStats` returns 100GB but we regularly query 70GB of underlying data, that's fine. We'll just set the `query.max-bytes=100GB`. Basically, base our idea of what the limit should be on the `IndexStats` response, not the amount of data we ended up querying.